### PR TITLE
fix(package-rules): matchCurrentVersion with null versioning

### LINKED ISSUE #24964: `lockFileMaintenance` and a `packageRule` with `matchCurrentVersion` cause Renovate crash
### Discussed in https://github.com/renovatebot/renovate/discussions/24958

<div type='discussions-op-text'>

<sup>Originally posted by **Badbond** October  2, 2023</sup>
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.152.0

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

When trying to upgrade Renovate from 35.141.3 to 35.159.7 (slim) in our self-hosted solution, we noticed that Renovate crashed in certain repositories with specific Renovate configuration. We were able to isolate the issue to release [35.152.0](https://togithub.com/renovatebot/renovate/releases/tag/35.152.0).

When combining `lockFileMaintenance.enabled: true` together with a `packageRule` that uses `matchCurrentVersion`, the `versioning` in [here](https://github.com/renovatebot/renovate/blob/6a2564dbeb93141a7f8b06fa29fdc61aa9a315a0/lib/util/package-rules/current-version.ts#L23) is `null`, which is invalid input for the Zod library [here](https://github.com/renovatebot/renovate/blob/6a2564dbeb93141a7f8b06fa29fdc61aa9a315a0/lib/modules/versioning/index.ts#L17). The logs attached show the error that is thrown.

https://github.com/Badbond/renovate-reproduction-case-lockfile-maintenance shows a minimal reproduction case that shows this behavior.

This issue is still apparent in 37.3.0 (latest at the time of testing). The issue is also apparent with the Mend Renovate GitHub App. In fact, the logs attached are those of the GitHub app run against the reproduction repository.

It furthermore seems a simple `null` to `''` conversion of `versioning` inside `modules/versioning/index.ts` fixes it, which seems to be in line with the intentions of the default for this argument. If you agree on this issue, would you like me to open a PR for this?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: branchifyUpgrades
ERROR: Repository has unknown error
{
  "err": {
    "message": "Schema error",
    "stack": "ZodError: Schema error\n    at Object.get error [as error] (/opt/containerbase/tools/renovate/37.0.3/node_modules/zod/lib/types.js:43:31)\n    at Object.get (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/modules/versioning/index.ts:23:25)\n    at CurrentVersionMatcher.matches (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/util/package-rules/current-version.ts:23:35)\n    at matcherOR (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/util/package-rules/utils.ts:19:27)\n    at matchesRule (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/util/package-rules/index.ts:17:30)\n    at applyPackageRules (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/util/package-rules/index.ts:74:9)\n    at flattenUpdates (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/repository/updates/flatten.ts:146:43)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at branchifyUpgrades (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/repository/updates/branchify.ts:21:19)\n    at lookup (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/repository/process/extract-update.ts:195:36)\n    at extractDependencies (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/repository/process/index.ts:161:11)\n    at Object.renovateRepository (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/repository/index.ts:62:9)\n    at attributes.repository (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/global/index.ts:184:11)\n    at start (/opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/workers/global/index.ts:169:7)\n    at /opt/containerbase/tools/renovate/37.0.3/node_modules/renovate/lib/renovate.ts:18:22",
    "issues": "Expected string, received null"
  }
}
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above</div>

### PR DESCRIPTION
**Pull request by rarkins**
Fixes #24964

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/package-rules/current-version.ts [hunk 1]
```diff
@@ -20,7 +20,7 @@ export class CurrentVersionMatcher extends Matcher {
     }
     const isUnconstrainedValue =
       !!lockedVersion && is.nullOrUndefined(currentValue);
-    const version = allVersioning.get(versioning);
+    const version = allVersioning.get(versioning ? versioning : undefined);
```
Comment (secustor) [inline lib/util/package-rules/current-version.ts:23]:
```suggestion
    if (is.nullOrUndefined(versioning)) {
      logger.debug(`No versioning found for package rule matchCurrentVersion '${matchCurrentVersion}'`);
      return null;
    }
    const version = allVersioning.get(versioning);
```

Why not simply returning if versioning is undefined?

Comment (viceice) [reply lib/util/package-rules/current-version.ts:23]: looks like a config error and we should also issue a warning?

**Review by viceice — COMMENTED** [review]
see comment

File: lib/util/package-rules/current-version.ts (hunk 1 above)
Comment (rarkins) [reply lib/util/package-rules/current-version.ts:23]: If it's undefined then it is defaulted to semver-coerced today. Null does not fall back though.

Returning false here would be wrong

Comment (secustor) [reply lib/util/package-rules/current-version.ts:23]: returning null would ignore that rule and falsify the result. 

IMO we should rather change this here: 
https://github.com/renovatebot/renovate/blob/c8fceec593a184cc7b858bb1186dd52c6a1206ed/lib/modules/versioning/index.ts#L19

```diff
- export function get(versioning = ''): VersioningApi {
+ export function get(versioning = defaultVersioning.id): VersioningApi { 
```

Comment (rarkins) [reply lib/util/package-rules/current-version.ts:23]: This change wouldn't coerce null, right? See my updated code

Comment (rarkins) [inline lib/util/package-rules/current-version.ts:23]:
```suggestion
    const version = allVersioning.get(versioning)
```

File: lib/util/package-rules/current-version.ts [hunk 2]
```diff
@@ -20,7 +20,7 @@ export class CurrentVersionMatcher extends Matcher {
     }
     const isUnconstrainedValue =
       !!lockedVersion && is.nullOrUndefined(currentValue);
-    const version = allVersioning.get(versioning);
+    const version = allVersioning.get(versioning)
```
Comment (rarkins) [inline lib/util/package-rules/current-version.ts:23]:
```suggestion
    const version = allVersioning.get(versioning);
```

File: lib/util/package-rules/current-version.spec.ts
Comment (secustor) [inline lib/util/package-rules/current-version.spec.ts:1]: 🤔  I would create a separate test case else we are testing two failure cases at once. 

That way it is getting harder to pinpoint regressions.

Comment (rarkins) [reply lib/util/package-rules/current-version.spec.ts:1]: Actually this is not an error case we're testing. It's a simple "false" response, and the versioning is irrelevant as regex is used

Comment (rarkins) [reply lib/util/package-rules/current-version.spec.ts:1]: Added a new test instead

**Review by secustor — APPROVED** [review]

**Pull request merged**